Konev

nolit1 opened the pull request.

nolit1 committed:
Сделаны все задачи из первого урока.

nolit1 committed:
Сделаны все задачи из первого урока и задача из второго.

maximpetrov commented:
recheck all

gagarski commented:
recheck all

gagarski commented on .gitignore:
```diff
@@ -1,11 +0,0 @@
-# Created by .ignore support plugin (hsz.mobi)
```
Зачем вы удалили .gitignore?
Из-за этого в ваши коммиты попадают изменения конфигурации среды разработки, которые не имеют отношения к решению задач.

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +60,31 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  * Вывести значение того же угла в радианах (например, 0.63256).
  */
-fun angleInRadian(grad: Int, min: Int, sec: Int): Double = TODO()
+fun angleInRadian(grad: Int, min: Int, sec: Int): Double = (grad+min/60.0+sec/3600.0)/57.2958
```
Ваша константа для деления задана недостаточно точно. Используйте константу  `PI` из класса `Math`.

nolit1 committed:
Сделаны все задачи из второго урока, кроме двух последних.

nolit1 committed:
испавлена задача из 2 урока

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +69,12 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+if (((kingX==rookX1)||(kingY==rookY1))&&(kingX==rookX2||(kingY==rookY2))) return(3)
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski replied:
[Здесь и везде] `return` - это оператор, а не функция. Скобки вокруг возвращаемого значения не нужны: `return 3`

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,24 +87,27 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
- */
-fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int  {
+    if (((kingX == rookX) || (kingY == rookY)) && ((kingX - bishopX) == (kingY - bishopY))) return(3) else
```
`else` в цепочке ветвлений принято переносить на новую строку вместе со следующим `if`:

``` kotlin
if (someCondition) someAction()
else if (someOtherCondition) someOtherAction()
else anotherAction()
```

или так:

``` kotlin
if (someCondition) 
    someAction()
else if (someOtherCondition) 
    someOtherAction()
else 
    anotherAction()
```

В таком случае среда разработки будет сама правильно форматировать код и вам не придётся двигать "лесенку" влево руками.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,24 +87,27 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+    /**
+     * Средняя
```
- В коммитах лучше избегать ненужных изменений, так как они усложняют объединение наборов изменений от разных разработчиков.

nolit1 committed:
исправлены задачи из 2 урока

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -88,9 +89,9 @@ else return(0)
 fun rookOrBishopThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
                           bishopX: Int, bishopY: Int): Int  {
-    if (((kingX == rookX) || (kingY == rookY)) && ((kingX - bishopX) == (kingY - bishopY))) return(3) else
```
[Здесь и далее] `return` - это оператор, а не функция. Скобки вокруг возвращаемого значения не нужны:  `return 3`

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -88,9 +89,9 @@ else return(0)
                           rookX: Int, rookY: Int,
                           bishopX: Int, bishopY: Int): Int  {
-    if (((kingX == rookX) || (kingY == rookY)) && ((kingX - bishopX) == (kingY - bishopY))) return(3) else
+    if (((kingX == rookX) || (kingY == rookY)) && ((Math.abs(kingX - bishopX)) == Math.abs(kingY - bishopY))) return(3) else
```
См. замечания к предыдущим коммитам про `else if` и форматирование.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -100,7 +101,14 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
      */
-    fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+    fun triangleKind(a: Double, b: Double, c: Double): Int {
+        if ((a>b+c || b>a+c || c>a+b)) return(-1)
```
См. замечания к предыдущим коммитам про `else if` и форматирование.

nolit1 committed:
доделан Lesson 2 task 1 до конца.

nolit1 committed:
исправлена последняя задача из lesson 2 task 1

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,8 +49,16 @@ fun ageDescription(age: Int): String = TODO()
+    val s1 = t1 * v1
+    val s2 = t2 * v2
+    val s3 = t3 * v3
+    val put: Double = ((s1 + s2 + s3) / 2.0)
```
- Постарайтесь избегать использования транслитерации в именах переменных.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,24 +87,41 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    else return(0)
+}
+    /**
+     * Простая
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,11 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String {
+    if ((age in 5..19) || age%10==0 || age%10>4 || (age%100 in 5..19))
+    return age.toString() + " лет"
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

nolit1 committed:
Доделан 2 lesson

gagarski commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +53,7 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean =
+        ((a<=r && b<=s) || (a<=s && b<=r))
```
Условие можно сильно упростить если заранее найти минимальные стороны кирпича.

gagarski commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +35,14 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
 fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean =
+        ((x1 + r1 == x2 + r2) && (x1 - r1 == x2 - r2) && (y1 + r1 == y2 + r2) && (y1 - r1 == y2 - r2)) ||
```
Вспомните уравнение окружности, попробуйте сделать чертёж на бумаге, условие можно записать **гораздо** проще.

gagarski replied:
актуально

nolit1 committed:
Сделано 6 задач из Lesson 3.

nolit1 committed:
Исправлены задачи из Lesson 3.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,67 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun digitNumber(n: Int): Int {
+    var count = 0
+    var number = n
+    do { count++
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,67 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int {
+    val k: Int
+    if (n > 2) k = ((fib(n - 1) + fib(n - 2))) else k = 1
```
- Рекурсивное решение напрашивается, исходя из определения чисел Фибоначчи, но попробуйте посчитать, к скольки вызовам функции `fib()` приведёт вызов `fib(10)`, `fib(40)`, `fib(55)`, сколько среди них вызовов с одинаковыми аргументами и решить задачу более оптимальным способом.

gagarski replied:
актуально

gagarski replied:
актуально

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,67 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun lcm(m: Int, n: Int): Int {
+    var own: Int = 1
```
Почитайте про алгоритмы поиска НОК. Есть более эффективные методы, чем полный перебор.

gagarski replied:
актуально

gagarski replied:
актуально

gagarski replied:
Вы уже реализовали вычисление НОД. Используйте его здесь.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,67 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun minDivisor(n: Int): Int {
+    var min: Int = 2
+
+    for (k in 2..n) {if (n%k == 0) break
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,67 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var max: Int = n-1
+    for (k in n-1 downTo 1) {if (n%max == 0) break
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,8 +128,17 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
-
+fun isCoPrime(m: Int, n: Int): Boolean {
+    if (m < 2 || n < 2) return false
```
Эту задачу можно решить в одну строку, применив одну из функций, написанную вами выше.

gagarski commented:
Пройдитесь пожалуйста по замечаниям и исправьте их.

nolit1 committed:
Решал задачи из 3 части

nolit1 committed:
Сделана сложная задача из 3 урока.

nolit1 committed:
3 урок, исправление старого и несколько новых задач

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -131,17 +179,27 @@ fun cos(x: Double, eps: Double): Double = TODO()
-
+fun isPalindrome(n: Int): Boolean {
+    val k: Int = revert(n)
+    if (n - k == 0) return true else return false
```
Здесь комбо из ненужных усложнений:
- Зачем проверять на `n - k == 0`, если можно на`n == k`
- `if` здесь не нужен. Можно написать просто `return n == k`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,17 +133,28 @@ fun maxDivisor(n: Int): Int = TODO()
-
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val nok = lcm(m, n).toDouble()
+    if ((m * n).toDouble() / nok == 1.0) return true
```
- Сравнивать результат операции в Double с другим числом типа Double - очень плохая идея
- Зачем делить их друг на друга?
- Зачем вообще преобразовывать их в `Double`?

Предыдущее решение хоть и было менее эффективным, но хотя бы без проблем, связанных со сравнением Double.

nolit1 replied:
fun isCoPrime(m: Int, n: Int): Boolean {
    val nok = lcm(m, n)
    if ((m \* n) / nok == 1) return true
    else return false
}

Можно ли так сделать? Я по формуле делал (НОК(a, b)=a·b:НОД(a, b)), в IDEA это работает

gagarski replied:
> Я по формуле делал (НОК(a, b)=a·b:НОД(a, b)), в IDEA это работает

Лучше примените эту формулу при вычислении НОК (я там писал вам замечание про неэффективность). Посчитайте НОД более эффективным алгоритмом (Гугл в помощь), а из него НОК. Заодно у вас появится функция НОД, которую вы сможете использовать здесь.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,17 +133,28 @@ fun maxDivisor(n: Int): Int = TODO()
+
+    if (m < 0 && n < 0) return false
+    for (k in m..n) {
+        if (Math.sqrt(k.toDouble()) * 10 % 10 == 0.0) break
```
- Зачем вы делаете `* 10 % 10`?
- Точное сравнение Double на равенство - это плохая идея

gagarski replied:
актуально

gagarski replied:
актуально

gagarski replied:
- Зачем вы делаете `* 10 % 10`?
- Точное сравнение Double на равенство - это плохая идея

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +215,14 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
  */
-fun squareSequenceDigit(n: Int): Int = TODO()
+fun squareSequenceDigit(n: Int): Int {
+    var string1: String = ""
```
Для больших `n` вы генерируете строку огромной длины. Решите задание без использования строк или хотя бы без хранения всей последовательности цифр.

nolit1 replied:
Постараюсь по-максимуму исправить все ошибки в заданиях

gagarski replied:
исправлено

nolit1 committed:
3 урок, исправление нескольих задач

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,17 +133,28 @@ fun maxDivisor(n: Int): Int = TODO()
-
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val nok = lcm(m, n)
+    if ((m * n) / nok == 1) return true
```
- Зачем вы делите? Из-за этого у вас ошибка (тут же целочисленное деление)
- Снова `if (...) true else false`

nolit1 replied:
Хорошо, сейчас посмотрю. Я просто хотел сейчас сдать сложную задачу ( И о, Господи, она работает), поэтому от этой задачи отошёл

nolit1 committed:
3 урок, сложная задача + несколько исправлений в других. Работает и хорошо

nolit1 committed:
4 урок

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,24 +87,41 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+     */
+    fun triangleKind(a: Double, b: Double, c: Double): Int {
+        if (a > b + c || b > a + c || c > a + b) return (-1)
+        if (sqr(a) > sqr(b) + sqr(c)) return (2) else if (sqr(a) == sqr(b) + sqr(c)) return (1)
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    var number = 0.0
+    for (i in 0..v.size - 1) {
```
[Здесь и далее] Если в теле цикла индекс вам не нужен, то можно обойтись без него:

``` kotlin
for (elem in v) {
    // elem - текущий элемент
}
```

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+fun abs(v: List<Double>): Double {
+    var number = 0.0
+    for (i in 0..v.size - 1) {
+        number += sqr(v[i].toString().toDouble())
```
Зачем `.toString().toDouble()`?!

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun mean(list: List<Double>): Double = TODO()
+fun mean(list: List<Double>): Double {
+    var sum = 0.0
+    for (i in 0..list.size - 1) {
```
См. замечания к `abs()`

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,43 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    val sum = mean(list)
+    if (list.size != 0) {
```
- См. замечания к `abs`
- `list.size == 0` здесь нет необходимости рассматривать как особый случай (и в вашем цикле и в `for (elem in list)` вы просто не попадёте в тело цикла, если длина списка равна нулю)

gagarski commented:
Исправьте предыдыщие ошибки и замечания, прежде чем делать новые задания. Если будут вопросы и ответы по замечаниям - задавайте.

nolit1 commented:
Спасибо за замечания. Вчера вечером исправлял и оптимизировал задачи. Понял как делать все намного проще и короче

nolit1 commented:
Помогите пожалуйста, каким способом лучше решать эту задачу?
 fun accumulate(list: MutableList<Double>): MutableList<Double> (4 lesson)
Я просто много всего перепробовал, но либо получается слишком большой код, либо просто не работает

nolit1 committed:
Исправлял и оптимизировал старые задачи, делал 4 урок

nolit1 commented:
Задачу из 2 урока случайно изменил, вечером исправлю и попробую оптимизировать , как писали ранее

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +175,13 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
-fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+// не работает, буду исправлять
+fun accumulate(list: MutableList<Double>): MutableList<Double> {
+    for ((index, elem) in list.withIndex()) {
```
Выглядит почти правильно, но подумайте о краевых условиях (посмотрите на исключение, которое бросается). Возможно тут как раз тот случай, когда есть смысл перебирать индексы списка в цикле.

nolit1 replied:
Спасибо, посмотрю

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,33 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    return Math.abs(Math.sqrt((v.map { it * it }).sum()))
```
Вспомните примеры однострочных функций из первого урока. Они не только позволяют вам сэкономить около десятка символов на скобки и `return`, но и позволяют не писать тип возвращаемого значения у функции:

``` kotlin
fun abs(v: List<Double>) = Math.abs(Math.sqrt((v.map { it * it }).sum()))
```

nolit1 commented:
А можно как-то поменять порядок элементов в списке, то есть [1,4,5,6] -> [6,5,4,1]?

nolit1 committed:
Осталось сделать 5 задач из 4 урока.

nolit1 committed:
4 урок. Вопрос по задаче convertToString

nolit1 commented:
Задача convertToString. Я уверен, что этот список, что я сделал, можно сократить в одну строчку, по типу для 10-35 заменить на a-z, как - не знаю. По крайней мере оно работает.  И, да, я не знаю, почему список назвал "dogs", сделал, потом хотел заменить на что-то нормальное, в итоге забыл. На работу не влияет, но всё же.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +236,47 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+     */
+    for (i in 0..dogs.size - 1) {
+        when {
+            dogs[i] == 10 -> string1 += "a"
```
- Можно воспользоваться тем, что в любой кодировке латинские символы хранятся друг за другом, а также тем, что к символу можно прибавлять число: `'a' + 1 == 'b'` и т. д.
- Конкатенация строк в цикле довольно неэффективна: при каждом `+=` предыдущая строка копируется. Эффективнее будет использовать список, а затем вызвать на нём метод `joinToString`.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +236,47 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val dogs = convert(n, base).toMutableList()
```
Зачем `toMutableList`?

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +285,11 @@ fun convertToString(n: Int, base: Int): String = TODO()
-fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    val a = digits.map { it.toDouble() }.toList()
+    return polynom(a.reversed(), base.toDouble()).toInt()
```
- Старайтесь избегать операций с числами с плавающей точкой в целочисленных задачах, чтобы не столкнуться с ошибками округления.

gagarski replied:
Актуально.

gagarski replied:
- Старайтесь избегать операций с числами с плавающей точкой в целочисленных задачах, чтобы не столкнуться с ошибками округления.

gagarski replied:
Стоит отметить, что кроме указанной проблемы с Double, сам подход к решению очень хорош.

nolit1 replied:
Функция polynom требует, чтобы число было типа double. Лишь поэтому в данной задаче я сначала преобразую в double

gagarski replied:
Думаю, что за элегантность решения я готов закрыть глаза на проблемы с Double в данной задаче, но исходное замечание следует иметь в виду.

nolit1 committed:
Весь 4й урок, кроме последней задачи

nolit1 committed:
4 урок весь.

nolit1 committed:
4 урок последняя задача

nolit1 committed:
4 урок, последняя задача

nolit1 committed:
4 урок

nolit1 committed:
4 урок последняя задача, 5 урок первые 2 задачи

nolit1 committed:
5 урок

nolit1 committed:
5 lesson 4 задачи

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +270,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list = listOf<Int>()
+    for (i in 0..str.length - 1) {
+        if (str[i].hashCode() > 57) {
```
- Что, по-вашему, вовзращает `hashCode()` от `Char`?
- Что такое 57?

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +270,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var list = listOf<Int>()
+    for (i in 0..str.length - 1) {
+        if (str[i].hashCode() > 57) {
+            list += 9 + str[i].hashCode() - 96
```
Что такое 96?

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +288,20 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+fun roman(n: Int): String {
+    var number = n
+    var i = 0
+    val listDigit = listOf(1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1)
```
Лучше список из пар или Map, чем два отведльных списка, неявно связанных индексом.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +310,38 @@ fun roman(n: Int): String = TODO()
+    val listHundred = listOf("", "сто", "двести", "триста", "четыреста", "пятьсот", "шестьсот", "семьсот", "восемьсот", "девятьсот")
+    val listDes = listOf("", "десять", "двадцать", "тридцать", "сорок", "пятьдесят", "шестьдесят", "семьдесят", "восемьдесят", "девяносто")
+    val listFin = mutableListOf<String>()
+    if (number % 100 == 0) Double.NaN else {
```
Отформатируйте `if - else if - else` пожалуйста

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +141,15 @@ fun dateDigitToStr(digital: String): String = TODO()
  */
-fun flattenPhoneNumber(phone: String): String = TODO()
+fun flattenPhoneNumber(phone: String): String {
+    if (phone.contains(Regex("""[^-+\d)(\s]"""))) return "" else {
```
- Отформатируйте `if - else if - else`
- Вы используете Regex'ы для задач, которые можно решить проще и понятнее без них.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +189,26 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var divisor: Int
+    var number = n
+    while (number > 1) {
+        divisor = minDivisor(number)
```
Вызывая `minDivisor`, вы в каждой итерации ищете минимальный делитель, начиная с 2. Можно искать его начиная с текущего делителя.

nolit1 replied:
Я не понимаю, что вы хотите, чтобы я сделал.

gagarski replied:
`minDivisor` каждый раз ищет минимальный делитель перебором, начиная с 2. 
После того, как вы поделили, например, на 157, нет смысла перебирать числа с 2 до 156.

nolit1 committed:
2 урок доделан, все остальные замечания завтра исправлю.

nolit1 committed:
2 и 4 уроки

nolit1 committed:
Исправил все замечания, решил ещё некоторые задачи.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -166,4 +252,16 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+    var string1: String = ""
+    var count = 0
+    val k: Int
+    if (n > 2) {k = ((fib(n - 1) + fib(n - 2)))
```
А почему не просто `fib(n)`?!

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +237,17 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val list = convert(n, base).toMutableList()
```
Зачем `toMutableList()`, если вы его потом не изменяете?

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -1,6 +1,9 @@
+
 package lesson5.task1
 
+import com.sun.javafx.binding.StringFormatter
```
- Зачем этот импорт?
- Старайтесь не импортировать ничего из пакетов `com.sun.*` и `sun.*`: http://stackoverflow.com/questions/1834826/it-is-a-bad-practice-to-use-suns-proprietary-java-classes

gagarski commented on test/lesson5/task1/Tests.kt:
```diff
@@ -46,18 +48,19 @@ class Tests {
     @Test
     @Tag("Normal")
     fun bestLongJump() {
+        assertEquals(2147483647, bestLongJump("1417255715 - - 819359384 - % - 1 2147483647 % - 813219160 2147483647 - 2147483647 0 % - 2147483647 - - 1026951656 % - % - 0 1291512751 - - 1 - - - - 2147483647 % - - 100547281 1 1 - 1 1850831090 - - - % 1135767028 2147483647 - % - - 1170676926 - - - - % - 250964373 - 0 - - - 1699779960 1 - 1021668586 - 2147483647 % - 2147483647 129207787 % - - - 2147 ...
```
Поэкспериментируйте с последовательностями из нескольких пробелов в разных местах. Скорее всего, Гитхаб потерял их из тестовой строки.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun minDivisor(n: Int): Int {
+    var min: Int = 2
+
+    for (k in 2..n) {
```
Можно ли уменьшить верхнюю границу перебора? А ещё сильнее, чем `n / 2`?

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var max: Int = n - 1
```
Можно ли уменьшить верхнюю границу перебора? А ещё сильнее, чем `n / 2`?

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +133,14 @@ fun maxDivisor(n: Int): Int = TODO()
  * Например, 25 и 49 взаимно простые, а 6 и 8 -- нет.
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
```
После исправления НОД/НОК можно решить задачу с помощью них.

gagarski reviewed (commented):
Просмотрите внимательно замечания к 3 уроку [тут](https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/55/files). Там ещё осталось кое-что.

nolit1 committed:
исправления + 7 урок

gagarski commented:
6 и 7 урок пока не смотрел потому что всё ещё есть неисправленные замечания к ранним урокам.

nolit1 commented:
Написал 6 и 7 чтобы получить зачет. Замечания буду исправлять, времени нет свободного просто.

nolit1 commented:
Все замечания, конечно, читаю и принимаю во внимание

nolit1 commented:
А почему у меня написано это?:
Author: Nolit1 [kj2806@mail.ru]

Owner: nolit1 []

Total: 21 / 57

Example: 9 / 9
Trivial: 1 / 1
Easy: 3 / 8
Normal: 6 / 18
Hard: 2 / 17
Impossible: 0 / 4

У меня ведь засчитаны намного больше задач. Откуда эти цифры?

gagarski commented:
> У меня ведь засчитаны намного больше задач. Откуда эти цифры?

Это итог по задачам, упомянутым в том комментарии (уроки 5-7).

gagarski commented:
recheck all

gagarski commented:
Верните пример `isPrime` в исходное состояние пожалуйста. Из-за ваших изменений не перепроверить все задания из-за таймаута :(

nolit1 committed:
3 урок, isPrime

nolit1 committed:
Исправил:
3 урок замечания minDivisor и maxDivisor исправил
4 урок замечание ConvertToString исправил

Не исправил:
4 урок Decimal не исправил, ибо не очень понимаю как. Комментарий про эту задачу оставил на GitHub
5 урок BestLongJump пока не смог исправить, чуть позже придумаю решение

Вроде бы все замечания...

В коммите будет первый урок ещё, лишь потому что там писал задачу на паре, там ничего не изменилось.

gagarski commented:
`isPrime` всё ещё изменён вами: смотрите [изменённые файлы](https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/55/files)

nolit1 commented:
Я правильно понимаю, так должно быть?
fun isPrime(n: Int): Boolean {
      if (n < 2) return false
     for (m in 2..Math.sqrt(n.toDouble()).toInt()) {
          if (n % m == 0) return false
      }
      return true

Я просто даже не помню как менял эту функцию...

nolit1 committed:
IsPrime. Надеюсь так было

nolit1 committed:
Все замечания исправил, ура.
В 5 уроке заменил на более рациональное и простое решение
В 3 уроке использовал НОД для isCoPrime, как и говорили

Не исправил только 1 задачу из 4 урока. Подскажите, как можно оставить мой алгоритм, но избавиться от double?

nolit1 commented:
Я сейчас исправил баги и некоторые моменты изменил совсем. Теперь всё работает, проверял неоднократно. Можете ли посмотреть и, если правильно, то засчитать работу?

fun main(args: Array<String>){
    println(exam(7))
    println(exam(9))
    println(exam(35))
}

fun exam(n: Int): MutableList<Int> {
    val result2 = mutableListOf<Int>()
    val list = mutableListOf(1,1)
    while(list[list.size - 1] - 1 < n){
        list.add(list[list.size - 1] + list[list.size - 2])
    }
    for (i in  list.size - 1 downTo 0) {
        result2.add(list[i])
        if (result2.sum() > n)
            result2[ result2.size - 1] = 0
    }

    if (result2.sum() != n)
        throw IllegalArgumentException()
    for (i in 0..result2.size - 1) {
        if (result2[i] != 0)
            result2[i] = 1
    }
    return result2.reversed().toMutableList()
}

gagarski commented:
recheck all

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +163,20 @@ fun flattenPhoneNumber(phone: String): String = TODO()
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+            val parts = jumps.split(" ", "%", "-").toMutableList()
+            for (i in 0..parts.size-1) {
```
Посмотрите в сторону функции `parts.removeAll {}` (ту, что принимает функцию-предикат как аргумент).

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +163,20 @@ fun flattenPhoneNumber(phone: String): String = TODO()
  * При нарушении формата входной строки или при отсутствии в ней чисел, вернуть -1.
  */
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
```
Фомратирование.

gagarski replied:
fixed

gagarski commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,61 @@ interface Matrix<E> {
+    }
+
+    override fun toString(): String {
+        val sb = StringBuilder()
```
Протестируйте ваш `toString`и сделайте так, чтобы вовзвращаемое им значение было читаемо человеком.

gagarski replied:
done

nolit1 committed:
Исправил все замечания, решил 1 задачу из 8

gagarski commented on src/lesson8/task1/Files.kt:
```diff
@@ -54,7 +54,14 @@ fun alignFile(inputName: String, lineLength: Int, outputName: String) {
  */
 fun countSubstrings(inputName: String, substrings: List<String>): Map<String, Int> {
-    TODO()
+    val result = mutableMapOf<String, Int>()
```
Форматирование

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -137,7 +173,29 @@ fun isLatinSquare(matrix: Matrix<Int>): Boolean = TODO()
-fun sumNeighbours(matrix: Matrix<Int>): Matrix<Int> = TODO()
+fun sumNeighbours(matrix: Matrix<Int>): Matrix<Int> {
+    val M = createMatrix(matrix.height, matrix.width, 0)
+    if (matrix.height == 1 && matrix.width == 1) return M
```
Нужно ли рассмотрение данного особого случая?

gagarski commented on src/lesson8/task1/Files.kt:
```diff
@@ -54,7 +54,14 @@ fun alignFile(inputName: String, lineLength: Int, outputName: String) {
+        val textResult = File(inputName).readText()
+        for (string in substrings) {
+                var count = 0
+                if (string.toLowerCase() in textResult.toLowerCase()) count += textResult.toLowerCase().split(string.toLowerCase()).size - 1
```
Вы вызываете `textResult.toLowerCase()` 2*substrings.size раз. Вызовите его вне цикла.

nolit1 commented:
У меня произошли проблемы с idea. В итоге я загрузил файл со своим проектом с git hub. Вроде бы всё стало хорошо, как минимум оно начало работать. Но пропали кнопки тестов (зелёные треугольники) и не подсвечиваются никакие ошибки (синтаксис и т.п.) Скажите, пожалуйста, что можно сделать?

gagarski commented:
После импорта Идея выдаёт сообщение "Non-managed pom.xml file found" (оно всплывает внизу после импорта, либо его можно найти в Event Log). Нужно нажать "Add as Maven Project".

После этого возможно появится сообщение "Project SDK is not defined", нужно выбрать SDK для проекта.

Если это не поможет, завтра посмотрим.

nolit1 committed:
исправлены замечания, сделана ещё 1 задача из 8

gagarski commented:
recheck all

gagarski commented:
Обратите внимание на падающий тест в  `lesson6.task2/square`.

Скорее всего вы не обрабатываете ошибки со слишком короткими строками.

nolit1 commented:
Хорошо, завтра исправлю эту программу, спасибо

gagarski commented:
Также, скорее всего, `countSubstrings` неправильно посчитает подстроки, если в файле есть пересекающиеся подстроки. Например, "бб" содержится в строке "ббб" два раза. Добавьте соответствующие тесты к данной функции и исправьте её, если она работает неправильно.

Исправьте `square` и `countSubstrings` до начала занятия и вы получите оценку "отлично".

nolit1 commented:
Понял, спасибо, сегодня сделаю

nolit1 committed:
Исправлена задача из 6 урока. Убрал из тестов 1 задачу из 8 урока, чтобы позднее её доделать для себя. На оценку "отлично", у меня все условия (нужное количество задач) выполнены , как я понял.

gagarski commented:
Хотелось бы чтобы вы всё-таки решили `countSubstrings`

nolit1 commented:
Обязательно доработаю эту задачу, только с долгами сейчас разберусь...

gagarski commented:
Сделаете до конца занятия в четверг - получите "отлично" =)

nolit1 commented:
Хорошо. Скажите, а где нужно прописывать свой вариант для теста? Просто задача ведь берёт текст из файла. Или как лучше это сделать?

gagarski commented:
Вы можете создать новый входной файл в каталоге `input`. Дальше - смотрите примеры уже созданных тестов.

nolit1 committed:
Сделал тесты, убедился что программа не работает. Исправил, теперь всё хорошо.
